Sources/NIOTCPEchoClient/Client.swift

```diff
@@ -0,0 +1,117 @@
+#if swift(>=5.9)
+@_spi(AsyncChannel) import NIOCore
+@_spi(AsyncChannel) import NIOPosix
+
+@available(macOS 14, *)
+@main
+struct Client {
+    /// The host to connect to.
+    private let host: String
+    /// The port to connect to.
+    private let port: Int
+    /// The client's event loop group.
+    private let eventLoopGroup: MultiThreadedEventLoopGroup
+
+    static func main() async throws {
+        let client = Client(
+            host: "localhost",
+            port: 8765,
+            eventLoopGroup: .singleton
+        )
+        try await client.run()
+    }
+
+    /// This method sends a bunch of requests.
+    func run() async throws {
+        try await withThrowingTaskGroup(of: Void.self) { group in
+            for i in 0...20 {
+                group.addTask {
+                    try await self.sendRequest(number: i)
+                }
+            }
+
+            try await group.waitForAll()
+        }
+    }
+
+    private func sendRequest(number: Int) async throws {
+        let channel = try await ClientBootstrap(group: eventLoopGroup)
+            .channelInitializer { channel in
+                channel.eventLoop.makeCompletedFuture {
+                    // We are using two simple handlers here to frame our messages with "\n"
+                    try channel.pipeline.syncOperations.addHandler(ByteToMessageHandler(NewlineDelimiterCoder()))
+                    try channel.pipeline.syncOperations.addHandler(MessageToByteHandler(NewlineDelimiterCoder()))
+                }
+            }
+            .connect(
+                host: self.host,
+                port: self.port
+            ) { channel in
+                channel.eventLoop.makeCompletedFuture {
+                    try NIOAsyncChannel(
+                        synchronouslyWrapping: channel,
+                        configuration: .init(
+                            inboundType: String.self,
+                            outboundType: String.self
+                        )
+                    )
+                }
+            }
+
+        print("Connection(\(number)): Writing request")
+        try await channel.outboundWriter.write("Hello on connection \(number)")
+
+        for try await inboundData in channel.inboundStream {
+            print("Connection(\(number)): Received response (\(inboundData))")
+
+            // We only expect a single response so we can exit here
+            break
+        }
+    }
+}
+
+/// A simple newline based encoder and decoder.
+private final class NewlineDelimiterCoder: ByteToMessageDecoder, MessageToByteEncoder {
+    typealias InboundIn = ByteBuffer
+    typealias InboundOut = String
+
+    private let newLine = "\n".utf8.first!
+
+    init() {}
+
+    func decode(context: ChannelHandlerContext, buffer: inout ByteBuffer) throws -> DecodingState {
+        let readable = buffer.withUnsafeReadableBytes { $0.firstIndex(of: self.newLine) }
```

Comment:
could we please not use `*Unsafe*` in the examples. `buffer.readableBytesView.firstIndex(of: UInt8(ascii: "\n"))` gives you what you need.

Reply:
or probably something like

```
let readableBytes = buffer.readableBytesView

if let firstLine = readableBytes.firstIndex(of: UInt8(ascii: "\n")).map({ readableBytes[readableBytes.startIndex ..< $0] }) {
    buffer.moveReaderIndex(forwardBy: firstLine.count)
    context.fireChannelRead(self.wrapInboundOut(ByteBuffer(firstLine.dropLast() /* kill newline */)))
    return .continue
} else {
    return .needMoreData
}
```